### PR TITLE
feat: channels repo

### PR DESCRIPTION
**Pull request by yo-ru**

### CODE REVIEW

**Comment by yo-ru** [comment]
LGTM? Tested fine in my cases.

File: app/repositories/channels.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+
+async def fetch_one(
+    id: Optional[int] = None,
+    name: Optional[str] = None,
+) -> Optional[dict[str, Any]]:
+    """Fetch a single channel."""
+    if id is None and name is None:
+        raise ValueError("Must provide at least one parameter.")
+    query = f"""\
```
Comment (cmyui) [inline app/repositories/channels.py:70]:
```suggestion
        raise ValueError("Must provide at least one parameter.")

    query = f"""\
```

File: app/repositories/channels.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        SELECT {READ_PARAMS}
+          FROM channels
+         WHERE id = COALESCE(:id, id)
+            AND name = COALESCE(:name, name)
```
Comment (cmyui) [inline app/repositories/channels.py:74]:
```suggestion
           AND name = COALESCE(:name, name)
```

File: app/repositories/channels.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+) -> int:
+    if read_priv is None and write_priv is None and auto_join is None:
+        raise ValueError("Must provide at least one parameter.")
+
+    query = """\
+        SELECT COUNT(*) AS count
+          FROM channels
+         WHERE read_priv = COALESCE(:read_priv, read_priv)
+            AND write_priv = COALESCE(:write_priv, write_priv)
+            AND auto_join = COALESCE(:auto_join, auto_join)
```
Comment (cmyui) [inline app/repositories/channels.py:98]:
```suggestion
           AND write_priv = COALESCE(:write_priv, write_priv)
           AND auto_join = COALESCE(:auto_join, auto_join)
```

File: app/repositories/channels.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    page: Optional[int] = None,
+    page_size: Optional[int] = None,
+) -> list[dict[str, Any]]:
+    """Fetch multiple channels from the database."""
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM channels
+         WHERE read_priv = COALESCE(:read_priv, read_priv)
+            AND write_priv = COALESCE(:write_priv, write_priv)
+            AND auto_join = COALESCE(:auto_join, auto_join)
```
Comment (cmyui) [inline app/repositories/channels.py:124]:
```suggestion
           AND write_priv = COALESCE(:write_priv, write_priv)
           AND auto_join = COALESCE(:auto_join, auto_join)
```

File: app/repositories/channels.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        """
+        params["limit"] = page_size
+        params["offset"] = (page - 1) * page_size
+
+    recs = await app.state.services.database.fetch_all(query, params)
+    return [dict(rec) for rec in recs]
+
+
+async def update(
+    name: str,
```
Comment (cmyui) [inline app/repositories/channels.py:145]: this should be by id

File: app/repositories/channels.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    """
+    params = {
+        "name": name,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    return dict(rec) if rec is not None else None
+
+
+async def delete(
+    name: str,
```
Comment (cmyui) [inline app/repositories/channels.py:182]: should also be by name

File: app/repositories/channels.py (hunk 5 above)
Comment (cmyui) [reply app/repositories/channels.py:145]: name support would be nice as well (channels are generally referenced by name) but for correctness let's just keep by id for now

File: app/repositories/channels.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    params = {
+        "name": name,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    if rec is None:
+        return None
+
+    query = """\
+        DELETE FROM channels
+         WHERE name = :name
```
Comment (cmyui) [inline app/repositories/channels.py:199]: this looks super bad, but it's what i'm doing in other places
```suggestion
        DELETE FROM channels
              WHERE name = :name
```

Comment (cmyui) [reply app/repositories/channels.py:199]: i might consider
```py
        DELETE
          FROM channels
         WHERE name = :name
```

File: app/repositories/channels.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+
+async def delete(
+    name: str,
+) -> Optional[dict[str, Any]]:
+    """Delete a channel from the database."""
+    query = f"""\
+        SELECT {READ_PARAMS}
+            FROM channels
+              WHERE name = :name
```
Comment (cmyui) [inline app/repositories/channels.py:188]:
```suggestion
        SELECT {READ_PARAMS}
          FROM channels
         WHERE name = :name
```

**Review by cmyui — APPROVED** [review]
+1 👍🏻

**Pull request merged**